add auth system

Graev opened the pull request:
Добавлена идентификация, аутентификация и авторизация

Graev committed:
add auth system

Добавлена идентификация, аутентификация и авторизация

aarefiev commented on controllers/cards.js:
```diff
@@ -1,54 +1,70 @@
-          res.status(500).send({ message: 'Произошла ошибка' });
+          res.status(500).send({ message: "Произошла ошибка" });
       }
       // res.status(500).send({ message: 'Произошла ошибка', err });
```
Нужно исправить: некорректная обработка ошибок.
Первое.
**Никогда** не стоит использовать тест в качестве условия обработки ошибки. Во первых, это затратно, так как сравнение строк происходит посимвольно. Во вторых, проверка не гарантирует работоспособность в будущем, так как любое изменение текста ломает проверку полностью.

Второе.
Валидация полей на существование должна производиться **до** сохранения. Смысл такой, что если поля не переданы, то и нет смысла пытаться сохранить это дело в базу.

Третье.
HTTP статус 404 означает, что ресурс **не найден**. Нужно возвращать 400, потому что переданы некорректные данные. А 500 ошибка верна - она значает, что произошла ошибка на сервере.

aarefiev replied:
Можно лучше: заменить `switch/case` на объект.
Хотя `switch/case` позволяет избавиться от нескольких `if`, но намного лучше использовать объект. В этом случае кода будет меньше.

aarefiev commented on controllers/cards.js:
```diff
@@ -1,54 +1,70 @@
+        ? 404
+        : 500;
+      const message =
+        statusCode === 404 ? "Неверно указан ID карточки" : "Произошла ошибка";
```
Нужно исправить: некорректная обработка ошибок.
Проблематика этого подхода указана в комментарии для функции `createCard`.

Понятно, что нет удобной возможности отредактировать какие данные будет возвращать база, при срабатывании ошибки. Но если разобрать что должно произойти при попытке удаления карточки, то эта проверка вообще становится ненужной. 

По заданию при попытке удаления карточки нужно сделать следующее:
1. Проверить доступы на удаление. По заданию только пользователь, который создал карточку может её удалить. Для это нужно получить карточку из базы и понять можно ли её удалить:
- если ошибка - стоит вернуть 400 и сообщение, что данные не верны,
- если карточка не найдена - стоит вернуть 404 и сообщение что карточка не найдена (точный текст сообщения можно взять из 12 практики),
- если карточка получена, то сравнить `owner`. Если `owner` соответствует, то идти к пункту 2, иначе - стоит вернуть 403 и сообщение, что нет доступа на удаление карточки.
2. Попытаться удалить карточку. Если удаление бросает исключение - значит происходит ошибка на сервере, потому что этап валидации данных и доступов уже пройден. В этом случае стоит вернуть 500 и сообщение, что произошла ошибка.

aarefiev commented on controllers/users.js:
```diff
@@ -1,64 +1,99 @@
+      const message =
+        statusCode === 404
+          ? "Неверно указан ID пользователя"
+          : "Произошла ошибка";
```
Нужно исправить: некорректная обработка ошибки.
Детали указаны для функции `deleteCard`.

aarefiev commented on controllers/users.js:
```diff
@@ -1,64 +1,99 @@
+  const { email, password, name, about, avatar } = req.body;
+  bcrypt.hash(password, 10).then(hash => {
+    password = hash;
+    User.create({ email, password, name, about, avatar })
```
Можно лучше: модифицировать переменную при необходимости.
Можно просто указать ключ, тогда не будет необходимости менять значение переменной:
```
User.create({ email, password: hash, name, about, avatar })
```

aarefiev commented on controllers/users.js:
```diff
@@ -1,64 +1,99 @@
+  bcrypt.hash(password, 10).then(hash => {
+    password = hash;
+    User.create({ email, password, name, about, avatar })
+      .then(user => res.status(201).send({ data: user }))
```
Так как не удалось запустить сервер, но возможно критическая ошибка существует. При сохранении пользователя ни в коем случае нельзя передавать скрытые данные обратно в целях безопасности. Возможно, в `user` содержится `password`, а его нельзя передавать в ответе.

aarefiev commented on controllers/users.js:
```diff
@@ -1,64 +1,99 @@
+          default:
+            res.status(500).send({ message: "Произошла ошибка" });
+        }
+        // res.status(500).send({ message: 'Произошла ошибка', err });
```
Нужно исправить: некорректная обработка ошибок.
Детали описаны для функции `createCard`.

aarefiev commented on controllers/users.js:
```diff
@@ -1,64 +1,99 @@
+      res.cookie("jwt", token, {
+        maxAge: 604800000,
+        httpOnly: true,
+        sameSite: true
```
Отлично!: обезопасим куку.

aarefiev commented on middlewares/auth.js:
```diff
@@ -0,0 +1,20 @@
+    playload = jwt.verify(token, process.env.JWT_SECRET);
+  } catch (err) {
+    return res.status(401).send({ message: "Необходима авторизация" });
+  }
```
Можно лучше: использовать асинхронные функции.
Модуль `jsonwebtoken` предоставляет асинхронные функции. Асинхронные функции лучше, так как не блокируют очередь выполнения.

aarefiev reviewed (commented):
Хорошо.

Проверить работоспособность не получилось. Сервер падает при обращении к любому ресурсу:
```bash
TypeError: Cannot read property 'jwt' of undefined
```

В коде обнаружил несколько критических ошибок, которые нужно исправить.

Нужно исправить:
1. Сервер падает при обращении к любому ресурсу.
2. Есть ошибки линтинга. Их достоточно много, поэтому привожу количество и команду, которая их запускает.
```bash
Project-on-Express$ npx eslint --ext .js .
...
✖ 128 problems (126 errors, 2 warnings)
```
Ошибки маркированные **error** обязательны к исправлению.
3. Некорректная обработка ошибок.

Можно лучше:
1. Подключить `helmet` (https://expressjs.com/ru/advanced/best-practice-security.html).
2. Cекретный ключ для выпуска jwt токена и для верификации жестко прописан в разных местах кода, а не реквайрится из одного места.
3. В разных файлах есть повторяющийся код. такое лучше выносить в отдельные модули.

Graev committed:
Error correction

Исправлена ошибка с падением сервера
Добавлен helmet
Исправлены ошибки линта
Оптимизированна проверка данных для создания и удаления пользователей/карточек

aarefiev commented on app.js:
```diff
@@ -1,33 +1,38 @@
 });
 
+app.post('/signin', login);
+app.post('/singup', createUser);
```
Нужно исправить: некорректное имя роута регистрации.
По заданию ожидается `/signup`.

aarefiev commented on models/users.js:
```diff
@@ -1,7 +1,23 @@
+  password: {
+    type: String,
+    required: true,
+    unique: true,
```
Нужно исправить: некорректная схема поля `password`.
Пароль - это дополнительная информация о пользователе. Основная - это адрес электронной почты. По паролю нельзя явно идентифицировать пользователя, поэтому пароль не может быть уникальным полем.

aarefiev commented on controllers/users.js:
```diff
@@ -1,64 +1,118 @@
+module.exports.login = (req, res) => {
+  const { email, password } = req.body;
+
+  User.findUserByCredentials({ email, password })
```
Нужно исправить: при попытке авторизации возвращается 401 и ошибка.
Ошибка:
```js
{ CastError: Cast to string failed for value "{ email: 'username4@mail.ru', password: 'password123456' }" at path "email" for model "user" }
```
Метод ожидает аргументы в виде списка, а не объекта:
https://github.com/Graev/Project-on-Express/blob/2dc01210b69d0e8514dc3b10074219a620e5870e/models/users.js#L44

aarefiev commented on controllers/users.js:
```diff
@@ -1,64 +1,118 @@
+        maxAge: 604800000,
+        httpOnly: true,
+        sameSite: true,
+      });
```
Нужно исправить: при попытке авторизации запрос не завершается.
В `express` метод [`.cookie()`](https://expressjs.com/ru/api.html#res.cookie) не отправляет ответ и закрывает соединение. Для этого есть метод [`.end()`](https://expressjs.com/ru/api.html#res.end). Либо можно использовать метод [`.send()`](https://expressjs.com/ru/api.html#res.send) для отправки тела ответа (при этом соединение будет закрыто).

aarefiev commented on middlewares/auth.js:
```diff
@@ -0,0 +1,29 @@
+    );
+  } catch (err) {
+    return res.status(401).send({ message: 'Необходима авторизация' });
+  }
```
Нужно исправить: роуты закрыты после авторизации.
Проблема в том, что при запросе на роут авторизации, токен будет установлен в куки пользователя:
https://github.com/Graev/Project-on-Express/blob/2dc01210b69d0e8514dc3b10074219a620e5870e/controllers/users.js#L85-L89

А мидлвара не может определиться откуда использовать токен:
- может из кук?
https://github.com/Graev/Project-on-Express/blob/2dc01210b69d0e8514dc3b10074219a620e5870e/middlewares/auth.js#L7
- может из заголовка `Authorization`?
https://github.com/Graev/Project-on-Express/blob/2dc01210b69d0e8514dc3b10074219a620e5870e/middlewares/auth.js#L16

Да из заголовка, потому что переменная `token` далее используется при проверке:
https://github.com/Graev/Project-on-Express/blob/2dc01210b69d0e8514dc3b10074219a620e5870e/middlewares/auth.js#L19-L22

Нужно заменить код, чтобы токен был получен из кук. Например, так:
```js
let playload;
try {
  playload = jwt.verify(
    req.cookie.jwt,
    'ea4d7d87d193658e6c08c5500e0e984b3102a6cf3a77f4678fff714eb75b023f'
  );
} catch (err) {
  return res.status(401).send({ message: 'Необходима авторизация' });
}
```

**Также не забыть удалить лишний код.**

aarefiev commented on controllers/users.js:
```diff
@@ -1,64 +1,118 @@
+    .then(user => {
+      const token = jwt.sign(
+        { _id: user._id },
+        'ea4d7d87d193658e6c08c5500e0e984b3102a6cf3a77f4678fff714eb75b023f',
```
Можно лучше: подключать секретный ключ из файла.
Если оставить секретный ключ в коде, то возникает несколько очевидных проблем:
- изменение ключа требует изменений всех файлов, где он используется,
- нельзя использовать отдельные ключи для разработки и продакшена 

Лучше создать отдельный файл (например, `config.js`) и в нём уже заполнять те конфигурационные данные, которые будут использоваться в проекте (например, секретный ключ, порт сервера или адрес для подключения к базе данных). Далее этот файл подключать в тех файлах, где необходимо получить значения. Пример файла:
```js
module.exports.JWT_TOKEN = 'secret key';
```

aarefiev reviewed (changes requested):
Очень хорошо. Огромный плюс за то, что были произведены изменения по комментариям из секции **Можно лучше**.

К сожалению, дальше части авторизации не удалось зайти. Поэтому ту часть не удалось запустить и проверить. Ниже я представил краткий список ошибок и улучшений, развернутые комментарии представлены под секцией **Можно лучше** и непосредственно в коде.

**Нужно исправить:**
1. Некорректное имя роута регистрации.
2. При попытке авторизации возвращается 401 и ошибка.
3. При попытке авторизации запрос не завершается.
4. Роуты закрыты после авторизации.
5. Некорректная схема поля `password`.

**Можно лучше**:
1. Оформить README.md.
2. Подключать секретный ключ из файла.

---

**Развернутый комментарий к улучшению "Оформить README.md"**:

Статья о том как написать хороший ридми: [ссылка](https://medium.com/astrolabe/%D0%BA%D0%B0%D0%BA-%D0%BD%D0%B0%D0%BF%D0%B8%D1%81%D0%B0%D1%82%D1%8C-%D1%85%D0%BE%D1%80%D0%BE%D1%88%D0%B8%D0%B9-readme-159f88076b26).
Официальный гайд github по оформлению: [официальная ссылка](https://help.github.com/en/articles/basic-writing-and-formatting-syntax), [он же, на русском](https://github.com/sandino/Markdown-Cheatsheet/blob/master/README.md).

Graev committed:
correct error and update README

Исправлено:
1) Имя роута для регистрации
2) добавил .end() для req.cookie
3)схема для поля password

Добавил подключение данных из файла config.
Дописан README.

aarefiev commented on middlewares/auth.js:
```diff
@@ -0,0 +1,15 @@
+module.exports = (req, res, next) => {
+  let playload;
+  try {
+    playload = jwt.verify(req.cookie.jwt, JWT_TOKEN);
```
Нужно исправить: после авторизации при попытке зайти на защищенный роут возвращается статус 401.
Здесь ошибка:
```bash
TypeError: Cannot read property 'jwt' of undefined
```

Вместо `req.cookie` должно быть `req.cookies`.

aarefiev commented on controllers/cards.js:
```diff
@@ -1,54 +1,64 @@
       if (!card) {
         return res.status(404).send({ message: 'Карточка не найдена' });
       }
+      if (!(card.owner._id === req.user._id)) {
```
Можно лучше: использовать `.equals()` при сравнении.
В библиотеке `mongoose` `card` - это экземпляр класса `Document`. У этого класса есть отдельный метод, который отвечает за проверку на равенство - это метод [`.equals()`](https://mongoosejs.com/docs/api.html#document_Document-equals). Его лучше использовать, потому что в этой практике поле `owner` - это объект, а `req.user._id` - это строка. И чтобы их сравнить нужно приводить типы. Если использовать метод библиотеки, то приведение типов делать не нужно.

aarefiev commented on controllers/cards.js:
```diff
@@ -1,54 +1,64 @@
 };
 
 module.exports.deleteCard = (req, res) => {
   Card.findByIdAndDelete(req.params.cardId)
```
Нужно исправить: карточка удаляется без проверки пользователя.
Метод `Card.findByIdAndDelete()` производит поиск карточки в базе и если она есть - удаляет. В задании, перед тем как удалить нужно удостовериться:
- есть ли карточка в базе - если нет - вернуть 404,
- есть ли у текущего пользователя доступ на удаление - если нет - вернуть 403

После того, как проверки выше пройдены - удаляем карточку. В самом начале нужно использовать метод [`Card.findById()`](https://mongoosejs.com/docs/api.html#model_Model.findById).

aarefiev commented on controllers/cards.js:
```diff
@@ -1,54 +1,64 @@
+    { $addToSet: { likes: req.user._id } },
+    { new: true }
+  )
+    .then(card => res.send({ data: card.likes }))
```
Нужно исправить: при попытке поставить лайк на несуществующую карточку возвращается статус 500.
Здесь ошибка:
```bash
TypeError: Cannot read property 'likes' of null
```
Метод `Card.findByIdAndUpdate()` производит поиск карточки в базе и если её нет - возвращает `Promise<null>`. У `null` нет свойства `likes`. Поэтому здесь нужно добавить проверку на значение и если карточки нет, то вернуть 404.

aarefiev commented on controllers/cards.js:
```diff
@@ -1,54 +1,64 @@
+    { $pull: { likes: req.user._id } },
+    { new: true }
+  )
+    .then(card => res.send({ data: card.likes }))
```
Нужно исправить: при попытке удалить лайк с несуществующей карточки возвращается статус 500.
Здесь ошибка:
```bash
TypeError: Cannot read property 'likes' of null
```
Метод `Card.findByIdAndUpdate()` производит поиск карточки в базе и если её нет - возвращает `Promise<null>`. У `null` нет свойства `likes`. Поэтому здесь нужно добавить проверку на значение и если карточки нет, то вернуть 404.

aarefiev commented on app.js:
```diff
@@ -1,33 +1,38 @@
+const { login, createUser } = require('./controllers/users');
+const auth = require('./middlewares/auth');
+const { PORT = 3000, SERVER_URL } = require('./config');
 
```
Можно лучше: не переопределять значения, которые получены из конфигурационного файла.
К тому же новое значение соответствует старому (возникает вопрос **Зачем?**):
https://github.com/Graev/Project-on-Express/blob/e4024075ea4b370f84d1f4bdd0fbacc3ec607681/config.js#L3

Дополню. Отдельный конфигурационный файл сделан не просто так, а чтобы абстрагировать конфигурацию от управляющего кода. Допустим, если нужно выложить проект на какой-нибудь удаленный сервер, те сделать продакшен версию, то это легко решается внутри конфигурационного файла:
```js
const JWT_TOKEN =
  process.env.JWT_TOKEN || 'ea4d7d87d193658e6c08c5500e0e984b3102a6cf3a77f4678fff714eb75b023f';
const PORT = 3000;
const SERVER_URL = process.env.SERVER_URL || 'mongodb://localhost:27017/mestodb';

module.exports = { JWT_TOKEN, PORT, SERVER_URL };
```

_Лучше всё-таки переименовать `SERVER_URL` в `DATABASE_URL`, второй вариант более связан с предметной областью использования._

В управляющем коде ничего не изменится, потому что импортируются значения из конфигурационного файла.

aarefiev commented on README.md:
```diff
@@ -1,5 +1,8 @@
 # Project-on-Express
```
Можно лучше: добавить форматирование в README.md.
Сейчас описание выглядит очень скомкано, как большая портянка текста.
![image](https://user-images.githubusercontent.com/2168159/73336313-e8fc7280-428a-11ea-948c-dc6d90d8e6a6.png)

Синтаксис **Markdown** позволяет добавить форматирование. В текущем ридми написано практически всё, что нужно и особо нет ничего лишнего, но, я уверен, не хватает акцента на деталях. Например, можно разделить всё что там есть на несколько секций:
- о проекте,
- настройка и запуск,
- особенности (сюда можно отнести список используемых библиотек и их версии).

Для примера можно взять шаблон: https://gist.github.com/PurpleBooth/109311bb0361f32d87a2 .

aarefiev reviewed (changes requested):
Очень хорошо. Огромный плюс за то, что были произведены изменения по комментариям из секции **Можно лучше**.

В этот раз не удалось попасть на роуты, которые закрыты авторизацией. Мне пришлось локально в коде внести исправление и продолжить проверку. Рекомендую перед сдачей на ревью несколько раз проверить сервер на работу и попытаться выполнить те кейсы, которые обрабатываются в роутах. Как и в прошлйы раз ниже краткий список ошибок и улучшений, развернутые комментарии можно найти в коде.

**Нужно исправить:**
1. Карточка удаляется без проверки пользователя.
2. При попытке поставить лайк на несуществующую карточку возвращается статус 500.
3. При попытке удалить лайк с несуществующей карточки возвращается статус 500.
4. После авторизации при попытке зайти на защищенный роут возвращается статус 401.

**Можно лучше:**
1. Добавить форматирование в README.md.
2. Не переопределять значения, которые получены из конфигурационного файла.
3. Использовать .equals() при сравнении.

Graev committed:
Error correction

Исправил ошибки вызванные невнимательностью (cookie -> cookies)
Оформил README
Изменил способ сравнения id в удалении карточки
Добавил проверку на создателя карточки до вызова функции удаления
Добавил проверку на существования карточки при установке/снятии лайка

aarefiev reviewed (approved):
Все выявленные ошибки исправлены, также внесены изменения по улучшениям. Получилось здорово, принято :tada: :rocket: .
